src/scatter.h

```diff
@@ -33,222 +35,335 @@ template <class T = float> class FormFactorSphere {
-        std::vector<T> _I(int((qmax - qmin) / dq + 0.5));
-        // loop over all pairs of groups, then over particles
-        for (int k = 0; k < (int)groupList.size() - 1; k++)
-            for (int l = k + 1; l < (int)groupList.size(); l++)
-                for (auto i : *groupList.at(k))
-                    for (auto j : *groupList.at(l)) {
-                        T r = geo.vdist(p[i].pos, p[j].pos).norm();
-                        int cnt = 0;
-                        for (T q = qmin; q <= qmax; q += dq)
-                            _I.at(cnt++) += sin(q * r) / (q * r);
-                    }
-        int cnt = 0, N = p.size();
-        for (T q = qmin; q <= qmax; q += dq)
-            I[q] += 2. * _I.at(cnt++) / N + 1; // add to average I(q)
+    /**
+     * @return a map containing q (key) and average intensity (value)
+     */
+    std::map<double, double> getIntensity() {
+        std::map<double, double> averaged_intensity;
+        for (int m = 0; m < intensity.size(); ++m) {
+            const T average = intensity[m] / (sampling[m] != 0.0 ? sampling[m] : 1.0);
+            averaged_intensity.emplace(q_mesh(m), average);
+        }
+        return averaged_intensity;
+    }
+
+};
+
+
+/**
+ * A policy for collecting samples. To be used together with StructureFactor class templates.
+ *
+ * @tparam T float or double
+ */
+template <typename T> class SamplingPolicy {
+  public:
+    struct sampled_value {
+        T value;
+        T weight;
+    };
+    typedef std::map<T, sampled_value> TSampledValues;
+  private:
+    TSampledValues samples;
+    const T precision = 10000.0; //!< precision of the key for better binning
+
+  public:
+    std::map<double, double> getSampling() {
+        std::map<double, double> average;
+        for (auto &kv : samples) {
+            average.emplace(kv.first, kv.second.value / kv.second.weight);
+        }
+        return average;
     }
 
     /**
-     * @brief Save I(q) to disk
+     * @param key_approx is subject of rounding for better binning
+     * @param value
+     * @param weight
      */
-    void save(const std::string &filename) {
-        if (!I.empty()) {
-            std::ofstream f(filename.c_str());
-            if (f) {
-                for (auto &i : I) {
-                    if (not S.empty())
-                        f << i.first << " " << i.second / S[i.first] << "\n";
-                    else
-                        f << i.first << " " << i.second << "\n";
-                }
-            }
-        }
+    void addSampling(T key_approx, T value, T weight = 1.0) {
+        const T key = std::round(key_approx * precision) / precision; // round |q| for better binning
+        samples[key].value += value;
+        samples[key].weight += weight;
     }
 };
 
+
 /**
- * @brief Structor factor calculation using explicit q averaging
+ * @brief Structure factor calculation using explicit q averaging.
```

Comment:
remove punctuation

Reply:
changed to a proper English sentence instead